### PR TITLE
[14.0] shopfloor: checkout scan change lot (rebased)

### PR DESCRIPTION
**Pull request by sebalix**
Same PR than https://github.com/OCA/wms/pull/569/ but rebased on top of 14.0 to resolve conflicts (we still need the previous PR to not bring regressions as it is used as pending-merge).

Comments from @jbaudoux have been addressed too.

cc @simahawk

### CODE REVIEW

**Comment by simahawk** [comment]
@sebalix are you going to address JE's comments too?

**Comment by sebalix** [comment]
@simahawk @jbaudoux comments addressed, excepting the check on the length of pickings in `_select_document_from_package` as it was breaking tests, indeed the relevant response is generated by `_select_picking` afterwards, I preferred to not change too much things for now and risk to break something else.

File: shopfloor/services/checkout.py [hunk 1]
```diff
@@ -556,15 +559,58 @@ def _select_lines_from_packaging(self, picking, selection_lines, packaging, **kw
...
+                )
+                # If there's at least one product matching we are good to go.
+                # In any case, only the 1st line matching will be affected.
+                if lines_same_product:
+                    return self._response_for_select_line(
+                        picking,
+                        message=self.msg_store.lot_different_change(),
+                        need_confirm_lot=lot.name,
+                    )
+                    # TODO: add a msg saying the lot has been changed
```
Comment (jbaudoux) [inline shopfloor/services/checkout.py]: to drop or misplaced TODO

File: shopfloor/services/checkout.py [hunk 2]
```diff
@@ -556,15 +559,58 @@ def _select_lines_from_packaging(self, picking, selection_lines, packaging, **kw
...
+                        selection_lines, lot
+                    )
+                )
+                # If there's at least one product matching we are good to go.
+                # In any case, only the 1st line matching will be affected.
+                if lines_same_product:
+                    return self._response_for_select_line(
+                        picking,
+                        message=self.msg_store.lot_different_change(),
+                        need_confirm_lot=lot.name,
```
Comment (jbaudoux) [inline shopfloor/services/checkout.py]: This is a bit unsafe as you can have many lots with the same name as the name is only unique per product.
A string with "model,id" would be more secure.

Comment (jbaudoux) [reply shopfloor/services/checkout.py]: Or just lot.id as we enter this method when a lot is scanned

File: shopfloor/services/checkout.py [hunk 3]
```diff
@@ -557,15 +569,58 @@ def _select_lines_from_packaging(self, picking, selection_lines, packaging, **kw
...
+            message = change_package_lot.change_lot(
+                line, lot, response_ok_func, response_error_func
+            )
+            if message["message_type"] == "error":
+                return self._response_for_select_line(picking, message=message)
+            else:
+                lines = line
+                # Some lines have been recreated, refresh the recordset
+                # to avoid CacheMiss error
+                selection_lines = self._lines_to_pack(picking)
```
Comment (TDu) [inline shopfloor/services/checkout.py:625]:
```suggestion
                selection_lines = self._lines_to_pack(picking)
                selection_lines = self._picking_lines_by_lot(picking, selection_lines, lot)
```
So only the lines with the lot that changed are displayed in the next state ?

**Comment by simahawk** [comment]
@TDu @jbaudoux @sebalix what's left to be done here beside rebasing n squashing?

File: shopfloor/services/checkout.py (hunk 3 above)
Comment (TDu) [reply shopfloor/services/checkout.py:625]: This is not relevant anymore.
A change related to this #785 will any one cause changes here

**Comment by TDu** [comment]
> @TDu @jbaudoux @sebalix what's left to be done here beside rebasing n squashing?

For me nothing but it is closely related to #785  and will need to be fixed on relation

**Review by TDu — APPROVED** [review]

**Comment by jbaudoux** [comment]
@TDu @sebalix Can this be rebased & squashed ? I understand there was nothing left to be done

**Comment by TDu** [comment]
@jbaudoux

**Pull request merged**